c_api/posit/posit_c_api.cpp

```diff
@@ -45,604 +42,139 @@ void unmarshal128(sw::unum::bitblock<128>& raw, posit128_t& a) {
+	public:
+	static sw::unum::posit<nbits, es> op(
+		sw::unum::posit<nbits, es> a,
+		sw::unum::posit<nbits, es> b
+	);
+};
+#define OPERATION(name, impl) \
+	template<size_t nbits, size_t es> class name: operation<nbits,es> { \
+		public: static sw::unum::posit<nbits, es> \
+			op(sw::unum::posit<nbits, es> a, sw::unum::posit<nbits, es> b) impl \
+	}
+OPERATION(op_add, { return a + b; });
+OPERATION(op_sub, { return a - b; });
+OPERATION(op_mul, { return a * b; });
+OPERATION(op_div, { return a / b; });
+
+template<size_t _nbits, size_t _es, class positN_t, class convert> class capi {
+	public:
+	static constexpr size_t nbits = _nbits;
+	static constexpr size_t es = _es;
+	static constexpr positN_t positN = positN_t();
+	static constexpr convert conv = convert();
+
+	static void format(positN_t p, char* str) {
+		using namespace sw::unum;
+		posit<nbits, es> pa = convert::decode(p);
+		std::string s = posit_format(pa);
+		sprintf(str, "%s", s.c_str());
+	}
 
-// posit<128,4> multiplication
-posit128_t posit_mul128(posit128_t a, posit128_t b) {
-	using namespace sw::unum;
-	constexpr size_t nbits = 128;
-	constexpr size_t es = 4;
-	posit<nbits, es> pa, pb, presult;
-	bitblock<nbits> raw;
-	marshal128(a, raw);
-	pa.set(raw);
-	marshal128(b, raw);
-	pb.set(raw);
-	presult = pa * pb;
-	// marshall it back to a posit128_t
-	posit128_t result;
-	bitblock<nbits> x = presult.get();
-	unmarshal128(x, result);
-	return result;
-}
+	template<class out>
+	static out to(positN_t bits) {
+		using namespace sw::unum;
+		posit<nbits, es> pa = convert::decode(bits);
+		return static_cast<out>(pa);
+	}
 
-///////////////////////////////////////////////////////////////
-/////////        DIVISION
+	template<class in>
+	static positN_t from(in a) {
+		using namespace sw::unum;
+		posit<nbits, es> pa(a);
+		return convert::encode(pa);
+	}
 
-// posit<8,0> division
-posit8_t posit_div8(posit8_t a, posit8_t b) {
-	using namespace sw::unum;
-	constexpr size_t nbits = 8;
-	constexpr size_t es = 0;
-	posit<nbits, es> pa, pb, presult;
-	pa.set_raw_bits(a);
-	pb.set_raw_bits(b);
-	presult = pa / pb;
-	return (posit8_t)presult.encoding();
-}
+	template<class operation>
+	static positN_t op(positN_t a, positN_t b) {
+		using namespace sw::unum;
+		posit<nbits, es> pa = convert::decode(a);
+		posit<nbits, es> pb = convert::decode(b);
+		posit<nbits, es> res = operation::op(pa, pb);
+		return convert::encode(res);
+	}
 
-// posit<16,1> division
-posit16_t posit_div16(posit16_t a, posit16_t b) {
-	using namespace sw::unum;
-	constexpr size_t nbits = 16;
-	constexpr size_t es = 1;
-	posit<nbits, es> pa, pb, presult;
-	pa.set_raw_bits(a);
-	pb.set_raw_bits(b);
-	presult = pa / pb;
-	return (posit16_t)presult.encoding();
-}
+	template<class ocapi>
+	static positN_t fromp(decltype(ocapi::positN) p) {
+		using namespace sw::unum;
+		posit<ocapi::nbits, ocapi::es> inp = decltype(ocapi::conv)::decode(p);
+		// TODO: There must be a better way to do this
```

Comment:
This is wrong, this should be using a correct posit conversion call rather than cramming everything into double and back.